src/invidious/routes/api/manifest.cr

```diff
@@ -20,6 +20,10 @@ module Invidious::Routes::API::Manifest
       haltf env, status_code: 403
     end
 
+    if local && CONFIG.invidious_companion
+      return env.redirect "#{video.invidious_companion["baseUrl"].as_s}#{env.request.path}?#{env.request.query}"
+    end
```

Comment:
There are multiple reasons:
- One may configure a proxy in invidious_companion. And invidious companion is going to handle multiple proxies soon, so this can't be handled inside Invidious:
- DASH manifest generation from youtube.js include way more things than invidious. which includes multi-language audio support, see here for more: https://github.com/LuanRT/YouTube.js/blob/4e9c2a585bf84751dd4e3964f70fba284c8b8e38/src/utils/DashManifest.tsx#L68-L267
- If an external program relies on the dash api, it makes more sense to redirect this request to Invidious companion because it avoids making unnecessary process in invidious as Invidious companion is the one requesting the video streams from YouTube.

Reply:
> One may configure a proxy in invidious_companion. And invidious companion is going to handle multiple proxies soon, so this can't be handled inside Invidious

Unless this is done transparently to the end user (e.g with NGINX rules), the CSP header will need to reflect the use of these external proxies, and so Invidious will need to take care of it.

> DASH manifest generation from youtube.js include way more things than invidious. which includes multi-language audio support

But is this compatible with videoJS? Our DASH manifest is missing multiple formats because of that.

> If an external program relies on the dash api, it makes more sense to redirect this request to Invidious companion because it avoids making unnecessary process in invidious as Invidious companion is the one requesting the video streams from YouTube.

Invidious already requests and parses the whole /player response returned by the companion (if it wasn't already in cache) so I don't think this will represent much more processing.

> One may configure multiple invidious companion, we need to make sure that the video stream used is from the same public IP address that generated the video stream.

For an API user, all URLs should already have that companion as the host, and for a web user, all the relevant URLs should already be handled in `components/player.ecr`. Am I right?

If so, the only way to access invidiou's own `/latest_version` and manifest endpoints would be willingly by a malicious or careless API user, so broken streams are a non-problem here imo.

Reply:
First of all, I would like to say that I have been in your thinking situation. I had long thoughts about the advantages and the disadvantages of each implementation. This current implementation outweighs the one where Invidious handle almost all the logic and Invidious companion is just "dumb proxy".

Also, invidious companion is not entirely new, it's actually based on the ideas of invidious-stripped-down. A project that I have used for 2 years on yewtu.be in order to replace some functions of Invidious. For really improving the performance and user experience of my instance.

> Unless this is done transparently to the end user (e.g with NGINX rules), the CSP header will need to reflect the use of these external proxies, and so Invidious will need to take care of it.

Already does in https://github.com/iv-org/invidious/pull/4985/files#diff-a86ced8aa99e3403588538decc008851d3d33dfadbb00d392d5ec8b4696f7df4

> But is this compatible with videoJS? Our DASH manifest is missing multiple formats because of that.

Yes. I have used for 2 years on yewtu.be. I'm stripping out the codec that do not work: https://github.com/iv-org/invidious-companion/blob/master/src/routes/invidious_routes/dashManifest.ts#L41-L43

> > If an external program relies on the dash api, it makes more sense to redirect this request to Invidious companion because it avoids making unnecessary process in invidious as Invidious companion is the one requesting the video streams from YouTube.
> 
> Invidious already requests and parses the whole /player response returned by the companion (if it wasn't already in cache) so I don't think this will represent much more processing.

The benefit of handling latest_version and /api/manifest/dash/id/ is that you can more easily handle all the edge cases in invidious companion itself, no need to add more logic in Invidious. Especially in a time when YouTube is evolving rapidly, we want to keep the pace thanks to youtube.js.

> For an API user, all URLs should already have that companion as the host, and for a web user, all the relevant URLs should already be handled in `components/player.ecr`. Am I right?
> If so, the only way to access invidiou's own `/latest_version` and manifest endpoints would be willingly by a malicious or careless API user, so broken streams are a non-problem here imo.

My number one priority is users using Invidious directly. We can think about API users later. Our main user base uses our frontend.

It doesn't mean that developers using the API are forever forgotten. I have ideas but for the moment, if invidious companion is configured then the video streams given through the API may not always work, especially when "proxying" through the same public IP address is needed.

I'm sorry, but a day is only 24 hours and at the moment I'm a single developer trying to improve the overall usability of Invidious through youtube.js. I can't deal with all the cases at the same time. Happy to receive more help for contribution!

Reply:
> Already does in https://github.com/iv-org/invidious/pull/4985/files#diff-a86ced8aa99e3403588538decc008851d3d33dfadbb00d392d5ec8b4696f7df4

The problem is that it add the companion's domains to the CSP, but not the companion's proxies themselves.

> The benefit of handling latest_version and /api/manifest/dash/id/ is that you can more easily handle all the edge cases in invidious companion itself [...]
> [...]
> My number one priority is users using Invidious directly. We can think about API users later. Our main user base uses our frontend.

My question was about the necessity of adding this redirect logic on these endpoints, where they should never be called in the first place, as none of the URLs in the `/watch` page should lead to invidious in the fist place!

Reply:
> > Already does in [#4985 (files)](https://github.com/iv-org/invidious/pull/4985/files#diff-a86ced8aa99e3403588538decc008851d3d33dfadbb00d392d5ec8b4696f7df4)
> 
> The problem is that it add the companion's domains to the CSP, but not the companion's proxies themselves.

This adds the baseUrl: https://github.com/iv-org/invidious-companion/blob/master/config/default.toml#L5 which is generated by invidious companion. This domain is the same domain to which videojs will send either latest_version or dash manifest api requests or videoplayback request.

We are not adding the URL passing in here: https://github.com/iv-org/invidious/pull/4985/files#diff-b68cf7cb2cb2dbd2275444b03cdc238962e05b5ccff0b89ba34b1f9126f39187R71
```
 invidious_companion:
 - http://127.0.0.1:8282
```

> My question was about the necessity of adding this redirect logic on these endpoints, where they should never be called in the first place, as none of the URLs in the `/watch` page should lead to invidious in the first place!

You always have stuff that use these endpoints directly without going through the frontend. It's to make sure that we are using invidious companion all the time.

- download function/button in invidious
- bots
- apps that use latest_version or dash manifest api instead of the official api of invidious

It's to make sure that we avoid sending wrong requests to youtube servers and the IP get banned for that.

Reply:
> This adds in the CSP the baseUrl [...] which is given by invidious companion in the JSON reply

I got that ^^ But if the companion has say `proxy1.example.com` and `proxy2.example.com` set, those won't be added to the CSP. Unless this is completely transparent to invidious (= the companion forwards the request to the proxies itself, and invidious only ever connects to `http://127.0.0.1:8282`)

> It's to make sure that we avoid sending wrong requests to youtube servers and the IP get banned for that.

Fair enough!

Reply:
I have rewritten some part. Let me explain.

If you ENABLE invidious_companion then by default in Invidious:
- /api/manifest/dash/id/ is blocked
- /latest_version is blocked

But for the "downloads" feature, all requests will now be straight sent to latest_version endpoint of invidious companion.

I know this does not yet download the video file directly but the user can do it by clicking on the download feature in their browser and this works. It's being tracked in https://github.com/iv-org/invidious-companion/issues/13

In my opinion this is not a dealbreaker as the Invidious download function is somewhat not useful anymore since the removal of hd720. Maybe in the future we could look into adding ffmpeg combining video and audio on the fly, but right now I'm prioritizing giving back more stability to Invidious.